File: src/whoop_client.py
"""
WHOOP API Client for MCP Server
"""
import httpx
import json
import logging
from typing import Dict, Any, Optional, List
from datetime import datetime, timedelta

from config import (
    WHOOP_API_BASE,
    REQUEST_TIMEOUT,
    MAX_REQUESTS_PER_MINUTE,
    CACHE_STORAGE_PATH,
    CACHE_DURATION
)
from auth_manager import TokenManager

logger = logging.getLogger(__name__)

class WhoopClient:
    """WHOOP API client with caching and rate limiting"""
    
    def __init__(self):
        self.base_url = WHOOP_API_BASE
        self.token_manager = TokenManager()
        self.cache = {}
        self.request_count = 0
        self.request_window_start = datetime.now()
        
    def _get_headers(self) -> Dict[str, str]:
        """Get headers with valid access token"""
        access_token = self.token_manager.get_valid_access_token()
        if not access_token:
            raise Exception("No valid access token available")
        
        return {
            'Authorization': f'Bearer {access_token}',
            'Content-Type': 'application/json'
        }
    
    def _check_rate_limit(self) -> None:
        """Check if we're within rate limits"""
        now = datetime.now()
        
        # Reset counter if window has passed
        if (now - self.request_window_start).total_seconds() >= 60:
            self.request_count = 0
            self.request_window_start = now
        
        if self.request_count >= MAX_REQUESTS_PER_MINUTE:
            raise Exception("Rate limit exceeded. Please wait before making more requests.")
        
        self.request_count += 1
    
    def _get_cache_key(self, endpoint: str, params: Dict[str, Any] = None) -> str:
        """Generate cache key for endpoint and parameters"""
        if params:
            param_str = json.dumps(params, sort_keys=True)
            return f"{endpoint}:{param_str}"
        return endpoint
    
    def _get_from_cache(self, cache_key: str) -> Optional[Dict[str, Any]]:
        """Get data from cache if still valid"""
        if cache_key in self.cache:
            cached_data = self.cache[cache_key]
            cache_time = datetime.fromisoformat(cached_data['cached_at'])
            
            if (datetime.now() - cache_time).total_seconds() < CACHE_DURATION:
                logger.debug(f"Cache hit for {cache_key}")
                return cached_data['data']
            else:
                # Remove expired cache entry
                del self.cache[cache_key]
        
        return None
    
    def _save_to_cache(self, cache_key: str, data: Dict[str, Any]) -> None:
        """Save data to cache"""
        self.cache[cache_key] = {
            'data': data,
            'cached_at': datetime.now().isoformat()
        }
        logger.debug(f"Cached data for {cache_key}")
    
    async def _make_request(self, endpoint: str, params: Dict[str, Any] = None) -> Dict[str, Any]:
        """Make authenticated request to WHOOP API"""
        # Check rate limits
        self._check_rate_limit()
        
        # Check cache first
        cache_key = self._get_cache_key(endpoint, params)
        cached_data = self._get_from_cache(cache_key)
        if cached_data:
            return cached_data
        
        # Make API request
        url = f"{self.base_url}/{endpoint.lstrip('/')}"
        headers = self._get_headers()
        
        try:
            async with httpx.AsyncClient(timeout=REQUEST_TIMEOUT) as client:
                response = await client.get(url, headers=headers, params=params or {})
                
                if response.status_code == 200:
                    data = response.json()
                    # Cache successful responses
                    self._save_to_cache(cache_key, data)
                    return data
                elif response.status_code == 401:
                    # Token might be expired, clear cache and try once more
                    self.token_manager.clear_tokens()
                    raise Exception("Authentication failed. Please re-authorize your WHOOP account.")
                else:
                    raise Exception(f"API request failed with status {response.status_code}: {response.text}")
                    
        except httpx.TimeoutException:
            raise Exception("Request timed out. Please try again.")
        except Exception as e:
            logger.error(f"Request failed for {endpoint}: {e}")
            raise
    
    async def get_user_profile(self) -> Dict[str, Any]:
        """Get user profile information"""
        return await self._make_request("/user/profile/basic")
    
    async def get_workouts(self, start_date: str = None, end_date: str = None, limit: int = 25) -> Dict[str, Any]:
        """Get user workouts"""
        params = {'limit': limit}
        
        if start_date:
            params['start'] = start_date
        if end_date:
            params['end'] = end_date
            
        return await self._make_request("/activity/workout", params)
    
    async def get_recovery(self, start_date: str = None, end_date: str = None, limit: int = 25) -> Dict[str, Any]:
        """Get user recovery data"""
        params = {'limit': limit}
        
        if start_date:
            params['start'] = start_date
        if end_date:
            params['end'] = end_date
            
        return await self._make_request("/recovery", params)
    
    async def get_sleep(self, start_date: str = None, end_date: str = None, limit: int = 25) -> Dict[str, Any]:
        """Get user sleep data"""
        params = {'limit': limit}
        
        if start_date:
            params['start'] = start_date
        if end_date:
            params['end'] = end_date
            
        return await self._make_request("/activity/sleep", params)
    
    async def get_cycles(self, start_date: str = None, end_date: str = None, limit: int = 25) -> Dict[str, Any]:
        """Get user physiological cycles"""
        params = {'limit': limit}
        
        if start_date:
            params['start'] = start_date
        if end_date:
            params['end'] = end_date
            
        return await self._make_request("/cycle", params)
    
    def get_auth_status(self) -> Dict[str, Any]:
        """Get authentication status"""
        return self.token_manager.get_token_info()
    
    def clear_cache(self) -> None:
        """Clear all cached data"""
        self.cache.clear()
        logger.info("Cache cleared")

File: smithery/src/whoop-client.ts
/**
 * WHOOP API Client for Smithery MCP Server
 */

import {
  WhoopConfig,
  WhoopProfile,
  WhoopWorkout,
  WhoopRecovery,
  WhoopSleep,
  WhoopCycle,
  ApiResponse,
  CacheEntry,
  RateLimitState,
  WhoopApiError
} from './types.js';

export class WhoopClient {
  private config: WhoopConfig;
  private cache = new Map<string, CacheEntry<any>>();
  private rateLimitState: RateLimitState = {
    requestCount: 0,
    windowStart: Date.now()
  };

  constructor(config: WhoopConfig) {
    this.config = config;
  }

  /**
   * Check and enforce rate limiting
   */
  private checkRateLimit(): void {
    const now = Date.now();
    const windowDuration = 60 * 1000; // 1 minute in milliseconds

    // Reset counter if window has passed
    if (now - this.rateLimitState.windowStart >= windowDuration) {
      this.rateLimitState.requestCount = 0;
      this.rateLimitState.windowStart = now;
    }

    if (this.rateLimitState.requestCount >= this.config.rateLimitPerMinute) {
      throw new Error('Rate limit exceeded. Please wait before making more requests.');
    }

    this.rateLimitState.requestCount++;
  }

  /**
   * Generate cache key for endpoint and parameters
   */
  private getCacheKey(endpoint: string, params?: Record<string, any>): string {
    if (params) {
      const paramStr = JSON.stringify(params, Object.keys(params).sort());
      return `${endpoint}:${paramStr}`;
    }
    return endpoint;
  }

  /**
   * Get data from cache if still valid
   */
  private getFromCache<T>(cacheKey: string): T | null {
    const cached = this.cache.get(cacheKey);
    if (cached && Date.now() < cached.expiresAt) {
      return cached.data as T;
    }

    if (cached) {
      // Remove expired entry
      this.cache.delete(cacheKey);
    }

    return null;
  }

  /**
   * Save data to cache
   */
  private saveToCache<T>(cacheKey: string, data: T): void {
    const expiresAt = Date.now() + (this.config.cacheTimeout * 1000);
    this.cache.set(cacheKey, {
      data,
      timestamp: Date.now(),
      expiresAt
    });
  }

  /**
   * Make authenticated request to WHOOP API
   */
  private async makeRequest<T>(
    endpoint: string,
    params?: Record<string, any>
  ): Promise<T> {
    // Check rate limits
    this.checkRateLimit();

    // Check cache first
    const cacheKey = this.getCacheKey(endpoint, params);
    const cached = this.getFromCache<T>(cacheKey);
    if (cached) {
      return cached;
    }

    // Build URL
    const url = new URL(endpoint, this.config.baseUrl);
    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined && value !== null) {
          url.searchParams.append(key, String(value));
        }
      });
    }

    // Make request
    const response = await fetch(url.toString(), {
      method: 'GET',
      headers: {
        'Authorization': `Bearer ${this.config.accessToken}`,
        'Content-Type': 'application/json',
        'Accept': 'application/json'
      }
    });

    if (!response.ok) {
      const errorText = await response.text();
      let errorMessage = `API request failed with status ${response.status}`;

      try {
        const errorJson = JSON.parse(errorText) as WhoopApiError;
        errorMessage = errorJson.message || errorMessage;
      } catch {
        // Use the default error message if JSON parsing fails
      }

      if (response.status === 401) {
        errorMessage = 'Authentication failed. Please check your access token.';
      } else if (response.status === 429) {
        errorMessage = 'Rate limit exceeded. Please wait before making more requests.';
      }

      throw new Error(errorMessage);
    }

    const data = await response.json() as T;

    // Cache successful responses
    this.saveToCache(cacheKey, data);

    return data;
  }

  /**
   * Get user profile information
   */
  async getUserProfile(): Promise<WhoopProfile> {
    const response = await this.makeRequest<ApiResponse<WhoopProfile>>('/user/profile/basic');
    return response.data;
  }

  /**
   * Get user workouts
   */
  async getWorkouts(params?: {
    start?: string;
    end?: string;
    limit?: number;
  }): Promise<WhoopWorkout[]> {
    const response = await this.makeRequest<ApiResponse<WhoopWorkout[]>>(
      '/activity/workout',
      { limit: 25, ...params }
    );
    return response.data;
  }

  /**
   * Get user recovery data
   */
  async getRecovery(params?: {
    start?: string;
    end?: string;
    limit?: number;
  }): Promise<WhoopRecovery[]> {
    const response = await this.makeRequest<ApiResponse<WhoopRecovery[]>>(
      '/recovery',
      { limit: 25, ...params }
    );
    return response.data;
  }

  /**
   * Get user sleep data
   */
  async getSleep(params?: {
    start?: string;
    end?: string;
    limit?: number;
  }): Promise<WhoopSleep[]> {
    const response = await this.makeRequest<ApiResponse<WhoopSleep[]>>(
      '/activity/sleep',
      { limit: 25, ...params }
    );
    return response.data;
  }

  /**
   * Get user physiological cycles
   */
  async getCycles(params?: {
    start?: string;
    end?: string;
    limit?: number;
  }): Promise<WhoopCycle[]> {
    const response = await this.makeRequest<ApiResponse<WhoopCycle[]>>(
      '/cycle',
      { limit: 25, ...params }
    );
    return response.data;
  }

  /**
   * Get authentication status
   */
  getAuthStatus(): { status: string; hasToken: boolean } {
    return {
      status: this.config.accessToken ? 'valid' : 'missing',
      hasToken: Boolean(this.config.accessToken)
    };
  }

  /**
   * Clear all cached data
   */
  clearCache(): void {
    this.cache.clear();
  }

  /**
   * Get cache statistics
   */
  getCacheStats(): { size: number; entries: string[] } {
    return {
      size: this.cache.size,
      entries: Array.from(this.cache.keys())
    };
  }
}